experimental/web/sample_webgpu/main.c

```diff
@@ -279,6 +280,197 @@ static iree_status_t parse_inputs_into_call(
+  iree_buffer_map_userdata_t* userdata =
+      (iree_buffer_map_userdata_t*)userdata_ptr;
+  switch (map_status) {
+    case WGPUBufferMapAsyncStatus_Success:
+      break;
+    case WGPUBufferMapAsyncStatus_Error:
+      fprintf(stderr, "  buffer_map_sync_callback status: Error\n");
+      break;
+    case WGPUBufferMapAsyncStatus_DeviceLost:
+      fprintf(stderr, "  buffer_map_sync_callback status: DeviceLost\n");
+      break;
+    case WGPUBufferMapAsyncStatus_Unknown:
+    default:
+      fprintf(stderr, "  buffer_map_sync_callback status: Unknown\n");
+      break;
+  }
+
+  if (map_status != WGPUBufferMapAsyncStatus_Success) {
+    iree_allocator_free(iree_allocator_system(), userdata);
+    return;
+  }
+
+  iree_status_t status = iree_ok_status();
+
+  // TODO(scotttodd): bubble result(s) up to the caller (async + callback API)
+
+  iree_device_size_t data_offset = iree_hal_buffer_byte_offset(
+      iree_hal_buffer_view_buffer(userdata->buffer_view));
+  iree_device_size_t data_length =
+      iree_hal_buffer_view_byte_length(userdata->buffer_view);
+  WGPUBuffer buffer_handle =
+      iree_hal_webgpu_buffer_handle(userdata->mappable_buffer);
+
+  // For this sample we want to print arbitrary buffers, which is easiest
+  // using the |iree_hal_buffer_view_format| function. Internally, that
+  // function requires synchronous buffer mapping, so we'll first wrap the
+  // already (async) mapped GPU memory into a heap buffer. In a less general
+  // application (or one not requiring pretty logging like this), we could
+  // skip a few buffer copies and other data transformations here.
+
+  const void* data_ptr =
+      wgpuBufferGetConstMappedRange(buffer_handle, data_offset, data_length);
+
+  iree_hal_buffer_t* heap_buffer = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_heap_buffer_wrap(
+        userdata->mappable_buffer->device_allocator,
+        IREE_HAL_MEMORY_TYPE_HOST_LOCAL, IREE_HAL_MEMORY_ACCESS_READ,
+        IREE_HAL_BUFFER_USAGE_MAPPING, data_length,
+        iree_make_byte_span((void*)data_ptr, data_length),
+        (iree_hal_buffer_release_callback_t){
+            .fn = iree_webgpu_mapped_buffer_release,
+            .user_data = buffer_handle,
+        },
+        &heap_buffer);
+  }
+
+  // Copy the original buffer_view, backed by the mapped heap buffer instead.
+  iree_hal_buffer_view_t* buffer_view = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_view_create(
+        heap_buffer, iree_hal_buffer_view_shape_rank(userdata->buffer_view),
+        iree_hal_buffer_view_shape_dims(userdata->buffer_view),
+        iree_hal_buffer_view_element_type(userdata->buffer_view),
+        iree_hal_buffer_view_encoding_type(userdata->buffer_view),
+        iree_allocator_system(), &buffer_view);
+  }
+
+  if (iree_status_is_ok(status)) {
+    fprintf(stdout, "Call output:\n");
+    status = iree_hal_buffer_view_fprint(stdout, buffer_view,
```

Comment:
I'm not sure it's worth it (given this is sample code and such) but one thing we could do in the future is refactor the iree_hal_buffer_view_fprint & co methods to all bottom out on a base method that took the buffer view info as flattened params - then you'd not need to create the new wrapper view here. More code though and probably worth seeing if this is a pattern we'd want to use.